### ISSUE TITLE
torch.nn.functional.pairwise_distcance fail to get the right grad when tensor is big (Tensor[100, 42949673])

### ISSUE THREAD

**Issue by wangyuwen1999** [pytorch/pytorch#159750]
### 🐛 Describe the bug

<img width="749" height="422" alt="Image" src="https://github.com/user-attachments/assets/3b92bafc-de7c-49c8-b079-a8d0cb7c5b18" />
by formula, the grad of $x_i$ should be $sign(x_i) * grad\_out$, however we get the value is less than grad_out
<img width="986" height="510" alt="Image" src="https://github.com/user-attachments/assets/8b40b616-d4a4-4bfd-b5bd-60036a85c948" />
We use paddle as a comparison，the abs of $x_i$'s grad should be similar to the grad_out value.

the code we use is 
```
import paddle
import torch
import numpy
import math
import re
def torch_assert_accuracy( paddle_tensor, torch_tensor, atol=1e-2, rtol=1e-2):

    if not paddle_tensor.is_contiguous():
        paddle_tensor = paddle_tensor.contiguous()
    paddle_tensor = paddle_tensor.cpu().detach()

    if not torch_tensor.is_contiguous():
        torch_tensor = torch_tensor.contiguous()
    torch_tensor = torch_tensor.cpu().detach()

    paddle_dlpack = paddle.utils.dlpack.to_dlpack(paddle_tensor)  # type: ignore
    converted_paddle_tensor = torch.utils.dlpack.from_dlpack(paddle_dlpack)  # type: ignore

    def error_msg(msg):
        return (
            f"Not equal to tolerance rtol={rtol}, atol={atol}\n"
            f"{msg}\n"
            f"ACTUAL: (shape={converted_paddle_tensor.shape}, dtype={converted_paddle_tensor.dtype})\n"
            f"{converted_paddle_tensor}\n"
            f"DESIRED: (shape={torch_tensor.shape}, dtype={torch_tensor.dtype})\n"
            f"{torch_tensor}"
        )

    is_backward = True 

    try:
        torch.testing.assert_close(
            converted_paddle_tensor,
            torch_tensor,
            rtol=rtol,
            atol=atol,
            equal_nan=True,
            check_dtype=True,
            msg=error_msg,
        )
    except Exception as e:
        error_str = str(e)
        print(f"[accuracy error] backward {error_str}", flush=True)
        matches = re.findall(r'\((\d+),\s*(\d+)\)', error_str)
        indices = [tuple(map(int, match)) for match in matches]
        return indices


x = numpy.load("/data/wangyuwen/workspace/PaddleAPITest/tmp_numpy_tensor/0-[100, 42949673]-float16.npy")
y = numpy.load("/data/wangyuwen/workspace/PaddleAPITest/tmp_numpy_tensor/1-[100, 42949673]-float16.npy")
grad_out = numpy.load("/data/wangyuwen/workspace/PaddleAPITest/tmp_numpy_tensor/2-torch.Size([100])-float16.npz")['arr0']

p_x = paddle.to_tensor(x, stop_gradient=False)
p_y = paddle.to_tensor(y, stop_gradient=False)
p_go = paddle.to_tensor(grad_out, stop_gradient=False)
p_z = paddle.nn.functional.pairwise_distance(p_x, p_y, math.inf, 1e-06, False, None)
p_g = paddle.grad(p_z, [p_x, p_y], p_go, allow_unused=True)

t_x = torch.tensor(x, requires_grad=True)
t_y = torch.tensor(y, requires_grad=True)
t_go = torch.tensor(grad_out, requires_grad=False)
t_z = torch.nn.functional.pairwise_distance(t_x, t_y, math.inf, 1e-06, False)
t_g = torch.autograd.grad(t_z, [t_x, t_y], t_go)

indexs = torch_assert_accuracy(p_g[0], t_g[0])

a,b = indexs[0]

print(f"mismatch in dx ({a},{b}) p_g: {p_g[0][a][b]} t_g: {t_g[0][a][b]} ground truth: {grad_out[a]}")
print(f"mismatch in dy ({a},{b}) p_g: {p_g[1][a][b]} t_g: {t_g[1][a][b]} ground truth: {grad_out[a]}")
```

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (GCC) 8.2.0
Clang version: 3.8.0 (tags/RELEASE_380/final)
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.10.18 (main, Jun  5 2025, 13:14:17) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.14.0_1-0-0-44-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 525.60.13
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              160
On-line CPU(s) list: 0-159
Thread(s) per core:  2
Core(s) per socket:  20
Socket(s):           4
NUMA node(s):        4
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
Stepping:            7
CPU MHz:             3200.000
CPU max MHz:         2501.0000
CPU min MHz:         1000.0000
BogoMIPS:            5000.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            28160K
NUMA node0 CPU(s):   0-19,80-99
NUMA node1 CPU(s):   20-39,100-119
NUMA node2 CPU(s):   40-59,120-139
NUMA node3 CPU(s):   60-79,140-159
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-nccl-cu11==2.21.5
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] torch==2.6.0+cu118
[pip3] torchaudio==2.6.0+cu118
[pip3] torchvision==0.21.0+cu118
[pip3] triton==3.2.0
[conda] numpy                       2.2.6            pypi_0           pypi
[conda] nvidia-cublas-cu11          11.11.3.6        pypi_0           pypi
[conda] nvidia-cuda-cupti-cu11      11.8.87          pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu11      11.8.89          pypi_0           pypi
[conda] nvidia-cuda-runtime-cu11    11.8.89          pypi_0           pypi
[conda] nvidia-cudnn-cu11           9.1.0.70         pypi_0           pypi
[conda] nvidia-cufft-cu11           10.9.0.58        pypi_0           pypi
[conda] nvidia-curand-cu11          10.3.0.86        pypi_0           pypi
[conda] nvidia-cusolver-cu11        11.4.1.48        pypi_0           pypi
[conda] nvidia-cusparse-cu11        11.7.5.86        pypi_0           pypi
[conda] nvidia-nccl-cu11            2.21.5           pypi_0           pypi
[conda] nvidia-nvtx-cu11            11.8.86          pypi_0           pypi
[conda] torch                       2.6.0+cu118      pypi_0           pypi
[conda] torchaudio                  2.6.0+cu118      pypi_0           pypi
[conda] torchvision                 0.21.0+cu118     pypi_0           pypi
[conda] triton                      3.2.0            pypi_0           pypi
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan

**Comment by albanD** [comment]
Hi,

Isn't this expected precision loss due to floating point inacuracies? https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html
